cupyx/scipy/sparse/_index.py

```diff
@@ -127,6 +127,214 @@ def _csr_row_index(rows, Ap, Aj, Ax):
+                        cur_insert_index =
+                            insert_indices[insert_row_start+insert_offset];
+                        cur_insert_value =
+                            insert_values[insert_row_start+insert_offset];
+                    }
+                }
+
+            } else if(input_offset < input_count) {
+                Bj[output_n] = cur_existing_index;
+                Bx[output_n] = cur_existing_value;
+
+                ++input_offset;
+                if(input_offset < input_count) {
+                    cur_existing_index = Aj[input_row_start+input_offset];
+                    cur_existing_value = Ax[input_row_start+input_offset];
+                }
+
+            } else {
+                    Bj[output_n] = cur_insert_index;
+                    Bx[output_n] = cur_insert_value;
+
+                    ++insert_offset;
+                    if(insert_offset < insert_count) {
+                        cur_insert_index =
+                            insert_indices[insert_row_start+insert_offset];
+                        cur_insert_value =
+                            insert_values[insert_row_start+insert_offset];
+                    }
+            }
+
+            output_n++;
+        }
+    ''', 'csr_copy_existing_indices_kern', no_return=True)
+
+
+# Create a filter mask based on the lowest value of order
+_unique_mask_kern = core.ElementwiseKernel(
+    '''raw I rows, raw I cols, raw I order''',
+    '''raw bool mask''',
+    """
+    I cur_row = rows[i];
+    I next_row = rows[i+1];
+
+    I cur_col = cols[i];
+    I next_col = cols[i+1];
+
+    I cur_order = order[i];
+    I next_order = order[i+1];
+
+    if(cur_row == next_row && cur_col == next_col) {
+        if(cur_order < next_order)
+            mask[i] = false;
+        else
+            mask[i+1] = false;
+    }
+    """, no_return=True
+)
+
+
+"""
+/*
+ * Determine the data offset at specific locations
+ *
+ * Input Arguments:
+ *   I  n_row         - number of rows in A
+ *   I  n_col         - number of columns in A
+ *   I  Ap[n_row+1]   - row pointer
+ *   I  Aj[nnz(A)]    - column indices
+ *   I  n_samples     - number of samples
+ *   I  Bi[N]         - sample rows
+ *   I  Bj[N]         - sample columns
+ *
+ * Output Arguments:
+ *   I  Bp[N]         - offsets into Aj; -1 if non-existent
+ *
+ * Return value:
+ *   1 if any sought entries are duplicated, in which case the
+ *   function has exited early; 0 otherwise.
+ *
+ * Note:
+ *   Output array Bp must be preallocated
+ *
+ *   Complexity: varies. See csr_sample_values
+ *
+ */
+"""
+_csr_sample_offsets_ker = core.ElementwiseKernel(
+    '''I n_row, I n_col, raw I Ap, raw I Aj, raw I Bi, raw I Bj''',
+    'raw I Bp, raw bool dupl', '''
+    const I j = Bi[i]; // sample row
+    const I k = Bj[i]; // sample column
+    const I row_start = Ap[j];
+    const I row_end   = Ap[j+1];
+    I offset = -1;
+    for(I jj = row_start; jj < row_end; jj++)
```

Comment:
Why not avoid this loop by using `sum_duplicated`?

Reply:
I tried to avoid sorting the entire array unless absolutely necessary (e.g., when a duplicate is actually found). One of the reasons for constructing the final array in sorted order during `insert_many` is that it was much less expensive than adding the insertions after the existing values and sorting.

Reply:
And I just realized that some of my more recent changes already assume sorted order, so it needs to be done anyways.

Reply:
The only way I can think of to avoid this loop is to convert the `indptr` to a COO row array, which would use more memory but would make the offset finding more parallel. Without knowing the offsets of the indices we are setting, it's not obvious to me how to set these values. Please let me know what I can do to fix this.

Reply:
I was able to remove the inner loop iterations, at least. When you get a moment, can you PTAL?

Reply:
How about creating a sparse array of [nnz of initial array] + [size of Bx] size, and then apply `sum_duplicated` to it? I think all loops in element-wise kernels.

Reply:
`sum_duplicates` may not be needed. It seems OK to just return the array of nnz of initial array + size of Bx size.